File: ros2/ros2_qos_examples_py/ros2_qos_examples_py/latest_frame_subscriber.py
"""Slow consumer (50 ms/frame) on a 60 Hz publisher.

With qos_profile_sensor_data the subscriber sees only the freshest frame each
callback — intermediate frames are dropped at the queue, not buffered.
"""

import time

import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from rclpy.time import Time
from sensor_msgs.msg import Image


class LatestFrameSubscriber(Node):
    def __init__(self) -> None:
        super().__init__('latest_frame_subscriber_py')
        self.sub = self.create_subscription(
            Image, 'camera/image', self._on_frame, qos_profile_sensor_data)

    def _on_frame(self, msg: Image) -> None:
        time.sleep(0.05)  # simulate slow processing
        marker = msg.data[0] if len(msg.data) else 0
        now = self.get_clock().now()
        stamp = Time.from_msg(msg.header.stamp)
        age_ms = (now - stamp).nanoseconds / 1e6
        self.get_logger().info(
            f'processed frame marker={marker}, age={age_ms:.1f} ms')


def main() -> None:
    rclpy.init()
    rclpy.spin(LatestFrameSubscriber())
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2/ros2_qos_examples_cpp/src/latest_frame_subscriber.cpp
// Slow consumer (50 ms/frame) on a 60 Hz publisher. With KEEP_LAST(5) +
// BEST_EFFORT, the subscriber sees only the freshest frame each callback —
// intermediate frames are dropped at the queue, not buffered.

#include <chrono>
#include <memory>
#include <thread>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"

using namespace std::chrono_literals;

class LatestFrameSubscriber : public rclcpp::Node
{
public:
  LatestFrameSubscriber() : Node("latest_frame_subscriber")
  {
    sub_ = create_subscription<sensor_msgs::msg::Image>(
      "camera/image",
      rclcpp::SensorDataQoS(),
      [this](sensor_msgs::msg::Image::ConstSharedPtr msg) { on_frame(msg); });
  }

private:
  void on_frame(sensor_msgs::msg::Image::ConstSharedPtr msg)
  {
    std::this_thread::sleep_for(50ms);

    const uint8_t marker = msg->data.empty() ? 0 : msg->data.front();
    const double age_ms = (now() - msg->header.stamp).seconds() * 1000.0;
    RCLCPP_INFO(get_logger(),
      "processed frame marker=%u, age=%.1f ms", marker, age_ms);
  }

  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr sub_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<LatestFrameSubscriber>());
  rclcpp::shutdown();
  return 0;
}
